### DISCUSSION TITLE
Blaze vs Edge

### DISCUSSION THREAD

**Discussion by harryadel** [meteor/blaze#442] [Ideas]
Hello, as I was going through the JS world on the internet I stumbled upon [Adonis](https://docs.adonisjs.com/guides/views/introduction#edge-vs-handlebars). It uses [Edge](https://edgejs.dev/docs/introduction) as its view layer. Edge one-ups Blaze/Handlebars in the ability to run JS directly in your view and upon further inspection it supports async/await using `let`:

![async_edge](https://github.com/meteor/blaze/assets/6644545/d8e6c207-5ea3-4eba-93f8-eb5473caa046)

It's a really promising library offering an enhanced version of Blaze/Handlerbars. Is there a way for us to utilize it or piggy-back off it instead of re-inventing the wheel? If we can find a backwards-compatible to integrate it into Meteor, it'd be a huge win since the Blaze community is small and would benefit a lot from having a well maintained library with a huge following.

Really interested to know your thoughts guys 
cc @StorytellerCZ @jankapunkt @DanielDornhardt

**Comment by jankapunkt** [comment]
This raises a fundamental question for the further direction of Blaze, as Blaze is strictly separating html template code from the JS part that is providing the actual contents. Many other libs, and also Edge, mix this code together and so does Svelte, Vue etc.

THe question is, if Blaze wants to go this same direction and if it's even capable of doing so

**Reply by harryadel** [reply]
If it's the question of capability that's holding you off then I assure you it's doable @zodern already illustrated how this can be done in Blaze in his "flare" series https://zodern.me/posts/flare-blaze-compiler/

**Comment by radekmie** [comment]
> Is there a way for us to utilize it or piggy-back off it instead of re-inventing the wheel?

If we'd like to utilize it, it'd require a viable `Tracker` integration. I don't think it's possible, since Edge is made with static rendering in mind, i.e., it always produces a _string_ and has nothing to do with the DOM. And that's really the hardest part.

As for taking something out of it... I don't know. We could think about some features to further extend Blaze, sure. But do you have anything particular in mind? Allowing arbitrary expressions sounds interesting, but I'm not sure if it's needed or if it wouldn't break the reactivity.

> If we can find a backwards-compatible to integrate it into Meteor, it'd be a huge win since the Blaze community is small and would benefit a lot from having a well maintained library with a huge following.

With all due respect to Edge creators and community, but... Is there a _huge following_? I knew about Adonis from some survey a few years back and never heard of Edge before this thread. It has 11k weekly downloads, which is substantial, sure. But if you ask me, I'd rather invest time into making the integration with one of the mainstream libraries better instead.

**Reply by harryadel** [reply]
>As for taking something out of it... I don't know. We could think about some features to further extend Blaze, sure. But do you have anything particular in mind? Allowing arbitrary expressions sounds interesting, but I'm not sure if it's needed or if it wouldn't break the reactivity.

Allowing arbitrary expressions seems really viable as I find myself constantly creating helpers to do basic computations using https://github.com/veliovgroup/Meteor-Template-helpers

>With all due respect to Edge creators and community, but... Is there a huge following? I knew about Adonis from some survey a few years back and never heard of Edge before this thread. It has 11k weekly downloads, which is substantial, sure. But if you ask me, I'd rather invest time into making the integration with one of the mainstream libraries better instead.

I understand where're your coming from as how we're not supposed to forcibly restrain ourselves to fit the narrative of another package/community and also the efficiency. You'd rather invest your time to attract 100 persons instead of 10. But let me remind you that every drop counts so hey if we can attract not 10 but a single developer I'm content. And I just liked how similar the syntax is between the two libs (Blaze & Edge).  It's a low hanging fruit, that's how I see it really. 

>integration with one of the mainstream libraries better instead

The big names like Vue/React or do you have something different in your mind?

**Reply by radekmie** [reply]
> Allowing arbitrary expressions seems really viable as I find myself constantly creating helpers to do basic computations using https://github.com/veliovgroup/Meteor-Template-helpers

Can you create a ticket for that? It'd be great to track it.

> You'd rather invest your time to attract 100 persons instead of 10. But let me remind you that every drop counts so hey if we can attract not 10 but a single developer I'm content.

Here's where we disagree. While I'd love to get more people to use Blaze, I don't think that spending _some_ amount of time to get _some_ amount of people in is sensible in all cases.

> And I just liked how similar the syntax is between the two libs (Blaze & Edge). It's a low hanging fruit, that's how I see it really.

As I said, it's not an easy thing. Edge was made for static rendering, Blaze for dynamic one. Even if we'd get the syntax to match, it'd still be a completely different technology underneath. And in that case, you could say we're using an even more popular Mustache syntax.

> The big names like Vue/React or do you have something different in your mind?

Yeah, those. But Svelte or SolidJS are fine too. Basically anything that has a decent community somehow compatible with Meteor. Adonis is a different framework so I don't expect anyone trying to mix it with Meteor.
